### PR TITLE
Features/enemies

### PR DESCRIPTION
**Pull request by VinayKashyap15**
Enemy classes added
Scriptable Objects included
-Nested Scriptable Object 
Require Component included
Collision check to destroy player or the enemy itself is present
New materials added for scriptable enemies

### CODE REVIEW

File: Assets/Scripts/Enemy/EnemyController.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            CreateModel(_enemyScriptableObject);           
+        }
+
+
+        private void SpawnEnemy(EnemyModel _enemyModel,Vector3 _position)
+        {
+           GameObject currentModel= GameObject.Instantiate(_enemyModel.GetEnemyModel());
+            currentModel.transform.position = _position;
+            currentModel.AddComponent<EnemyView>();
+            currentModel.GetComponentInChildren<Renderer>().sharedMaterial = _enemyModel.GetEnemyMaterial();
```
Comment (mayankgrover) [inline Assets/Scripts/Enemy/EnemyController.cs]: handling for materials etc should all be hidden inside the view, controller does not need to know about the inner details of the view

File: Assets/Scripts/Enemy/EnemyService.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        {
+            base.OnInitialize();
+
+            SpawnEnemyControllers();
+        }
+
+        private void SpawnEnemyControllers()
+        {
+            //Check if enemies to spawn is greater than scripatble objects provided
+            if(listOfEnemies.enemiesToSpawn>listOfEnemies.enemyList.Count)
```
Comment (mayankgrover) [inline Assets/Scripts/Enemy/EnemyService.cs]: game design wise its fine if you want each enemy to be unique but this doesn't let me spawn duplicate enemies of the same type so it imposes a design limitation

File: Assets/Scripts/Common/SingletonBase.cs [hunk 3]
```diff
@@ -20,19 +19,15 @@ private void Awake()
...
-                else
-                {
-                    Destroy(Instance);
-                }
+                instance = FindObjectOfType<T>();
+                DontDestroyOnLoad(instance);
+            }
+            else
+            {
+                Destroy(this.gameObject);
```
[the remark this answers is not in the document]
Comment (VinayKashyap15) [reply Assets/Scripts/Common/SingletonBase.cs]: done

File: Assets/Scripts/Enemy/EnemyView.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    [RequireComponent(typeof(Rigidbody))]
+    [RequireComponent(typeof(BoxCollider))]
+    public class EnemyView : MonoBehaviour
+    {
+        private EnemyController currentEnemyController;
+
+        //box collider default values if any idiot forgets to add a box collider. This will be not hardcoded if i expose this in scriptable object.
+        private void Awake()
+        {
+            gameObject.GetComponent<BoxCollider>().center = new Vector3(0, 1f, 0);
```
[the remark this answers is not in the document]
Comment (VinayKashyap15) [reply Assets/Scripts/Enemy/EnemyView.cs]: Done

File: Assets/Scripts/Enemy/EnemyService.cs (hunk 2 above)
Comment (VinayKashyap15) [reply Assets/Scripts/Enemy/EnemyService.cs]: resolved by adding another boolean isUnique

File: Assets/Scripts/Enemy/EnemyController.cs (hunk 1 above)
Comment (VinayKashyap15) [reply Assets/Scripts/Enemy/EnemyController.cs]: donee

**Pull request merged**